clustermesh: Implement MergeExternal* for experimental LB

joamaki opened the pull request:
Move the implementation from `experimental/adapters.go` into `pkg/clustermesh/service_merger.go` and remove from `adapters.go`. This more closely associates the logic with `pkg/clustermesh`.

Add ClusterID to backends so we can modify the backend associations with a service for specific clusterID (so that we won't remove backends added from other clusters).

Implement custom backend selection for ClusterMesh that implements ServiceAffinity.

Expose SelectBackends so it can be used to look up backends for a specific service and not just for specific frontend. Use this for pkg/ciliumenvoyconfig to look up backends to push to Envoy. 

For testing implement an in-memory kvstore which we can control from the clustermesh script tests.

joamaki committed:
clustermesh: Implement ServiceMerger for experimental LB

Implement the MergeExternal* methods for the experimental LB control-plane in
pkg/clustermesh. If experimental control-plane is not enabled use ServiceCache.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Remove Merge* methods from adapter

Now that external workloads has been removed and the MergeExternal* methods
are implemented in pkg/clustermesh we can remove the support for these from
the adapter.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
annotation: Move ClusterMesh annotations to pkg/annotation

Move ClusterMesh related annotations (Shared, IncludeExternal, ServiceAffinity)
to the pkg/annotation so they can be also used by pkg/loadbalancer/experimental.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
clustermesh: Script tests for ClusterService handling

Add a script test for testing that a ClusterService is properly
reflected into load-balancing tables.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
kvstore: Add in-memory kvstore backend for testing

Implement a simple in-memory kvstore backend module based on StateDB
for testing purposes. Not all operations are supported, nor is locking
supported.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

giorio94 commented on pkg/clustermesh/service_merger.go:
```diff
@@ -0,0 +1,121 @@
+		name,
+		source.ClusterMesh,
+		ClusterServiceToBackendParams(service)...,
+	)
```
IIUC, this (and symmetrically the deletion) would not work for more than two clusters (and yep, that's another area which is lacking proper testing). The reason being that `MergeExternalServiceUpdate` is called for each remote cluster, and the backends would end up being overwritten. The source should probably encode the cluster name somehow to avoid this problem.

joamaki replied:
Added `ClusterID` as argument to `SetBackends` and extended the tests to use 3 clusters.

giorio94 commented on pkg/clustermesh/service_merger.go:
```diff
@@ -0,0 +1,121 @@
+
+// MergeExternalServiceDelete implements k8s.ServiceCache.
+func (sm *expServiceMerger) MergeExternalServiceDelete(service *serviceStore.ClusterService, swg *lock.StoppableWaitGroup) {
+	if service.Cluster == option.Config.ClusterName {
```
Nit. You can retrieve `cmtypes.ClusterInfo` rather than accessing the global variable.

joamaki replied:
Ah great!

giorio94 commented on pkg/clustermesh/selectbackends.go:
```diff
@@ -0,0 +1,81 @@
+			useRemote = true
+		case experimental.ServiceAffinityLocal:
+			countLocal, countRemote := getCounts()
+			useLocal = countLocal > 0 || countRemote == 0
```
Nit. Could be simplified to `useLocal = true`, as you always want to use local endpoints, and if none are present it doesn't actually matter. Same below for the remote case.

joamaki replied:
Updated and added a test case.

joamaki committed:
experimental: Include ClusterID when manipulating backends

For ClusterMesh we may get backends for the same service from multiple clusters
and thus it is not enough to just use the Source for figuring out what backends
are orphans in SetBackends(). Add the ClusterID as an argument so we can
differentiate between the clusters.

Remove the unused ReleaseBackendsFromSource and ReleaseBackendsForService.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
clustermesh: Add service affinity test

Add a script test 'service-affinity.txtar' for testing the affinity rules.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

giorio94 commented on pkg/clustermesh/service_merger.go:
```diff
@@ -0,0 +1,123 @@
+		// prior backends.
+		sm.writer.DeleteBackendsOfService(txn, name, source.ClusterMesh)
+		return
+	}
```
Should be redundant now, as the [observer](https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/clustermesh/common/services.go#L193-L211) that eventually calls into this function already implements the same logic and ensures that this function is called only if `Shared` is true. Most likely, it was previously needed for external workloads.

joamaki replied:
Ah I see, I'll remove it then.

giorio94 commented on pkg/clustermesh/service_merger.go:
```diff
@@ -0,0 +1,123 @@
+		addrCluster, err := cmtypes.ParseAddrCluster(ipString)
+		if err != nil {
+			continue
+		}
```
Nit. This error should never occur, because the service is already validated after unmarshaling [here](https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/service/store/store.go#L150-L160).

giorio94 commented on pkg/loadbalancer/experimental/service_annotations.go:
```diff
@@ -0,0 +1,64 @@
+		return strings.ToLower(value) == "true"
+	}
+	return false
+}
```
Nit. It would probably make sense to eventually merge these functions with the already existing logic to parse these annotations (I assume calling directly into them would currently result in a circular import), just to prevent possible divergences in the future.

joamaki replied:
What do you mean with "existing logic"? The ones in `pkg/k8s`? Those are going to be deleted soon only leaving this one. Though if there's a need to parse these outside `*Service` then I guess we could consider keeping them around somewhere else. I didn't find a need for that with `git grep` though.

Importing `pkg/k8s` is fine though (already using it for `k8s.Endpoints`), so if you'd prefer that we'll keep the annotation parsing still in `pkg/k8s` I can do that. Or maybe these could be in `pkg/annotation` instead? Would be kind of natural as there we define the key of the annotation anyway.
E.g. we'd have functions to parse using `GetAnnotations() map[string]string` so it wouldn't pull in the k8s definitions and would work with both `v1.Service` and `experimental.Service` etc.

giorio94 replied:
> What do you mean with "existing logic"? The ones in pkg/k8s? Those are going to be deleted soon only leaving this one.

Yep, those ones. They are also used by the symmetric logic in the operator/clustermesh-apiserver, which may still rely on the old ServiceCache for a bit more, before being refactored as well.

joamaki replied:
Moved to pkg/annotation. This seems much nicer.

giorio94 commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -168,7 +168,7 @@ func (s *healthServer) controlLoop(ctx context.Context, health cell.Health) erro
 				if !needsServer {
 					wtxn := s.params.Writer.WriteTxn()
-					s.params.Writer.DeleteBackendsOfService(wtxn, healthServiceName, source.Local)
+					s.params.Writer.DeleteBackendsOfService(wtxn, healthServiceName, source.Local, 0)
```
Nit. Could define a constant to make the meaning more explicit.

joamaki replied:
Good idea. Added a constant. And also changed this so that there's `SetBackends` without clusterID and `SetBackendsOfCluster` with clusterID (since only ClusterMesh cares about giving a clusterID)

giorio94 commented on pkg/loadbalancer/experimental/backend.go:
```diff
@@ -256,10 +259,10 @@ func (be *Backend) release(name loadbalancer.ServiceName) (*Backend, bool) {
 }
 
-func (be *Backend) releasePerSource(name loadbalancer.ServiceName, source source.Source) (*Backend, bool) {
+func (be *Backend) releasePerSource(name loadbalancer.ServiceName, source source.Source, clusterID *uint32) (*Backend, bool) {
```
Nit. Does this need to be a pointer? It seems to be never passed nil.

joamaki replied:
Changed.

giorio94 commented on pkg/loadbalancer/experimental/writer.go:
```diff
@@ -396,16 +396,17 @@ func (w *Writer) UpsertBackends(txn WriteTxn, serviceName loadbalancer.ServiceNa
 // SetBackends sets the backends associated with a service. Existing backends from this source that
 // are associated with the service but are not given are released.
-func (w *Writer) SetBackends(txn WriteTxn, name loadbalancer.ServiceName, source source.Source, bes ...BackendParams) error {
+func (w *Writer) SetBackends(txn WriteTxn, name loadbalancer.ServiceName, source source.Source, clusterID uint32, bes ...BackendParams) error {
```
It could make sense to validate that the `clusterID` passed as parameter matches the one set in `BackendParams` (or maybe forcefully overwrite it?). Funky things could otherwise happen if they don't match.

joamaki replied:
I'll just forcefully overwrite it the same as with Source.

joamaki committed:
experimental: Expose SelectBackends for CiliumEnvoyConfig

Adapt and expose the SelectBackends function for use in CiliumEnvoyConfig controller
for picking which backends to send to Envoy. This allows for example ClusterMesh
to control which backends Envoy is aware of.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

giorio94 commented on pkg/clustermesh/service_merger.go:
```diff
@@ -0,0 +1,123 @@
+	if service.Cluster == option.Config.ClusterName {
+		// Ignore updates of own cluster and unshared services.
+		return
+	}
```
The clustermesh logic already [prevents connecting to the self cluster](https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/clustermesh/common/clustermesh.go#L168-L171), and it [validates](https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/clustermesh/clustermesh.go#L185-L189) that the services retrieved from a given cluster have the `Cluster` field set to the matching value. Hence, this check (and the symmetric one above) could now be dropped for simplicity.

giorio94 commented on pkg/loadbalancer/experimental/service_annotations.go:
```diff
@@ -0,0 +1,64 @@
+	return svc.Annotations
+}
+
+func (svc *Service) GetIncludeExternalAnnotation() bool {
```
Nit. There used to be a few tests in `pkg/k8s` for these functions. Fairly trivial, but it could make sense to port them here as well.

joamaki replied:
I ended up moving the annotations and tests to pkg/annotation. That seemed cleanest.

giorio94 commented on pkg/clustermesh/selectbackends.go:
```diff
@@ -0,0 +1,75 @@
+// injectSelectBackends overrides the load-balancing backend selection algorithm to implement
+// support for the ServiceAffinity and IncludeExternal annotations.
+func injectSelectBackends(cfg common.Config, expCfg experimental.Config, w *experimental.Writer) {
+	if cfg.ClusterMeshConfig == "" || !expCfg.EnableExperimentalLB {
```
The typical check to see if clustermesh is enabled would be: 

```go
func injectSelectBackends(cm *ClusterMesh, ...) {
	if cm == nil || ... {
```

Because the additional condition for it to be enabled is that the local ClusterID is different from zero.

giorio94 commented on pkg/loadbalancer/experimental/writer.go:
```diff
@@ -469,14 +480,18 @@ func (w *Writer) updateBackends(txn WriteTxn, serviceName loadbalancer.ServiceNa
-			} else {
-				_, _, err = w.bes.Insert(txn, be)
+				_, _, err = w.bes.Delete(txn, beNew)
+			} else if beNew != be {
```
Nit. This optimization could be moved to a separate commit.

joamaki replied:
Not an optimization. The logic changed in `releasePerSource` due to skipping backends that have different clusterID leading to it return `be` unchanged (which was not the case here before for this method).

giorio94 commented on pkg/ciliumenvoyconfig/exp_controller.go:
Nit. I would suggest removing the changes to `clustermesh/selectbackends.go` from this commit (`experimental: Expose SelectBackends for CiliumEnvoyConfig`), and squashing them into the commit that introduced that file, for better separation and ease of review.

giorio94 commented on pkg/clustermesh/selectbackends.go:
Nit. I'd suggest squashing these changes (last commit) into the commit that introduced this file as well, for ease of review.

giorio94 commented on pkg/clustermesh/selectbackends.go:
```diff
@@ -51,19 +52,21 @@ func (sb ClusterMeshSelectBackends) SelectBackends(bes iter.Seq2[experimental.Ba
+			// Always include the local backends even if they are unhealthy and
+			// only include (healthy) remote ones if there are no healthy local backends.
+			useLocal = true
+			useRemote = localBackends == 0 && remoteBackends > 0
```
Should we also include the remote backends if all the local ones are terminating? Otherwise I guess new connections would fail during that period of time.

joamaki replied:
Ah good point. Changed.

giorio94 commented on pkg/clustermesh/service_merger.go:
Nit. I would squash the changes in this file into the commit that initially introduced it, for consistency and ease of review.

joamaki replied:
Done

giorio94 commented on pkg/kvstore/memory.go:
```diff
@@ -0,0 +1,342 @@
+	_, existed, _ := c.table.Delete(wtxn, inMemoryObject{key: key})
+	if !existed {
+		return fmt.Errorf("key %q did not exist", key)
+	}
```
The `Delete` method does not return an error if the key does not exist: https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/kvstore/backend.go#L160-L161

giorio94 commented on pkg/kvstore/memory.go:
```diff
@@ -0,0 +1,342 @@
+	obj, _, found := c.table.Get(c.db.ReadTxn(), inMemoryKeyIndex.Query(key))
+	if !found {
+		return nil, fmt.Errorf("key %q not found", key)
+	}
```
`Get` does not return an error if the key does not exist: https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/kvstore/etcd.go#L1238-L1240

giorio94 commented on pkg/kvstore/memory.go:
```diff
@@ -0,0 +1,342 @@
+	changeIter, err := c.table.Changes(wtxn)
+	wtxn.Commit()
+	if err != nil {
+		panic("TODO")
```
Nit. The message could be extended a bit :grin:

joamaki replied:
heheh oops

giorio94 reviewed (commented):
Thanks! A couple more comments inline, mostly nits. A few are about slightly reordering the commits, just because now it is a tricky to review commit by commit IMO.

nathanjsweet reviewed (approved):
LGTM for sig-k8s

joamaki committed:
clustermesh: Custom backend selection for service affinity

Implement service affinity for ClusterMesh using a custom SelectBackends()
function that further filters the backends by:

* Only using local backends when service affinity set to local and there are
  healthy local backends. If local backends are unhealthy or all terminating
  then also remote cluster backends are used.
* Use only remote backends if service affinity set to remote and there are
  healthy remote backends. Use also local if no healthy remote backends available.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

mhofstetter reviewed (approved):
CEC integration LGTM - thanks!

giorio94 reviewed (approved):
Looks great, thanks a lot! :rocket:
